File: iosApp/iosApp/views/screens/countrydetail/CountryDetailScreen.swift
//
//  CountryDetailScreen.swift
//
//  Created by Daniele Baroncelli on 13/03/2021.
//

import SwiftUI
import shared

struct CountryDetailScreen: View {
    @ObservedObject var observableScreenState: ObservableScreenState
    
    init(
        observableScreenState: ObservableScreenState
    ) {
        self.observableScreenState = observableScreenState
    }
    
    var body: some View {
        VStack {
            let countryDetailState = self.observableScreenState.state as! CountryDetailState
            if countryDetailState.isLoading {
                LoadingScreen()
            } else {
                let data = countryDetailState.countryInfo
                ScrollView(.vertical) {
                    VStack(alignment: .leading, spacing: 5) {
                        DataElement(label: "total population", value: data.population)
                        DataElement(label: "    with first dose", value: data.firstDoses, percentage: data.firstDosesPerc)
                        DataElement(label: "    fully vaccinated", value: data.fullyVaccinated, percentage: data.fullyVaccinatedPerc)
                        Spacer().frame(height: 20)
                        if data.vaccinesList != nil  {
                            Text("Vaccines:").font(Font.callout.bold())
                            ForEach(data.vaccinesList!, id: \.self) { vaccine in
                                Text("   ‣ \(vaccine)").font(Font.callout)
                            }
                        }
                    }.padding(.all, 15).frame(maxWidth: .infinity, alignment: .leading)
                }
            }
        }
    }
}


struct DataElement: View {
    var label: String
    var value: String = ""
    var percentage: String = ""
    var body: some View {
        HStack {
            Text("\(label):").font(Font.callout.bold())
            Text(value).font(Font.callout)
            if percentage != "" {
                Text(" (\(percentage))").font(Font.callout)
            }
        }
    }
}

File: composeApp/src/commonMain/kotlin/screens/countrydetail/CountryDetailScreen.kt
package eu.baroncelli.dkmpsample.screens.countrydetail

import androidx.compose.foundation.layout.*
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import eu.baroncelli.dkmpsample.screens.LoadingScreen
import eu.baroncelli.dkmpsample.shared.viewmodel.screens.countrydetail.CountryDetailState

@Composable
fun CountryDetailScreen(
    countryDetailState: CountryDetailState,
) {

    if (countryDetailState.isLoading) {

        LoadingScreen()

    } else {
        val data = countryDetailState.countryInfo
        Column(modifier = Modifier.padding(10.dp)) {

            DataElement("total population", data.population)
            DataElement("   with first dose", data.firstDoses, data.firstDosesPerc)
            DataElement("   fully vaccinated", data.fullyVaccinated, data.fullyVaccinatedPerc)

            Spacer(modifier = Modifier.size(24.dp))

            if (data.vaccinesList!=null) {
                Text(text = "Vaccines:", style = MaterialTheme.typography.bodyLarge, fontWeight = FontWeight.Bold)
                for (vaccine in data.vaccinesList!!) {
                    Text(text = "   ‣ $vaccine", style = MaterialTheme.typography.bodyLarge)
                }
            }
        }

    }
}



@Composable
fun DataElement(label : String, value : String = "", percentage : String = "") {
    Row {
        Text(text = "$label: ", style = MaterialTheme.typography.bodyLarge, fontWeight = FontWeight.Bold)
        Text(text = value, style = MaterialTheme.typography.bodyLarge)
        if (percentage!="") {
            Text(text = " ($percentage)", style = MaterialTheme.typography.bodyLarge)
        }
    }
}